elements/lisk-chain/test/integration/state_store/save_diff.spec.ts

```diff
@@ -0,0 +1,256 @@
+import {
+	createFakeDefaultAccount,
+	defaultAccountSchema,
+	AccountAsset,
+} from '../../utils/account';
+import { DataAccess } from '../../../src/data_access';
+import { defaultBlockHeaderAssetSchema } from '../../utils/block';
+import { registeredTransactions } from '../../utils/registered_transactions';
+import { StateStore, Account } from '../../../src';
+import {
+	DB_KEY_DIFF_STATE,
+	DB_KEY_ACCOUNTS_ADDRESS,
+	DB_KEY_CHAIN_STATE,
+	DB_KEY_CONSENSUS_STATE,
+} from '../../../src/data_access/constants';
+import { stateDiffSchema } from '../../../src/schema';
+
+describe('stateStore.finalize.saveDiff', () => {
+	const defaultNetworkIdentifier = Buffer.from(
+		'93d00fe5be70d90e7ae247936a2e7d83b50809c79b73fa14285f02c842348b3e',
+	);
+	let db: KVStore;
+	let dataAccess: DataAccess;
+	let stateStore: StateStore;
+	let accounts: Account<AccountAsset>[];
+
+	beforeAll(() => {
+		const parentPath = path.join(__dirname, '../../tmp/diff');
+		fs.ensureDirSync(parentPath);
+		db = new KVStore(path.join(parentPath, '/test-diff.db'));
+		dataAccess = new DataAccess({
+			db,
+			accountSchema: defaultAccountSchema as any,
+			registeredBlockHeaders: {
+				0: defaultBlockHeaderAssetSchema,
+				2: defaultBlockHeaderAssetSchema,
+			},
+			registeredTransactions,
+			minBlockHeaderCache: 3,
+			maxBlockHeaderCache: 5,
+		});
+	});
+
+	beforeEach(() => {
+		stateStore = new StateStore(dataAccess, {
+			lastBlockHeaders: [],
+			networkIdentifier: defaultNetworkIdentifier,
+			lastBlockReward: BigInt(500000000),
+			defaultAsset: createFakeDefaultAccount().asset,
+		});
+
+		accounts = [
+			createFakeDefaultAccount({
+				address: Buffer.from('cc96c0a5db38b968f563e7af6fb435585c889111', 'hex'),
+				publicKey: Buffer.from(
+					'456efe283f25ea5bb21476b6dfb77cec4dbd33a4d1b5e60e4dc28e8e8b10fc4e',
+					'hex',
+				),
+				balance: BigInt(99),
+			}),
+			createFakeDefaultAccount({
+				address: Buffer.from('584dd8a902822a9469fb2911fcc14ed5fd98220d', 'hex'),
+				publicKey: Buffer.from(
+					'd468707933e4f24888dc1f00c8f84b2642c0edf3d694e2bb5daa7a0d87d18708',
+					'hex',
+				),
+				balance: BigInt('10000'),
+				keys: {
+					mandatoryKeys: [
+						Buffer.from(
+							'456efe283f25ea5bb21476b6dfb77cec4dbd33a4d1b5e60e4dc28e8e8b10fc4e',
+							'hex',
+						),
+					],
+					optionalKeys: [],
+					numberOfSignatures: 3,
+				},
+			}),
+		];
+	});
+
+	afterEach(async () => {
+		await db.clear();
+	});
+
+	describe('finalize', () => {
+		it('should save only account state changes diff', async () => {
+			// Arrange
+			accounts.map(account => stateStore.account.set(account.address, account));
+			const fakeHeight = '1';
+			const batch = db.batch();
+
+			// Act
+			stateStore.finalize(fakeHeight, batch);
+			await batch.write();
+			const diff = await db.get(`${DB_KEY_DIFF_STATE}:${fakeHeight}`);
+			const decodedDiff = codec.decode(stateDiffSchema, diff);
+
+			// Assert
+			expect(decodedDiff).toStrictEqual({
```

Comment:
Can we also include `undo` in this test, we haven't checked whether the diff generated can be used to rollback to previous state. so can we check `previousValue = undo(currentValue, decodedDiff)` if its too much else we could cover in another PR. Same for chain and consensus state tests